libs/utils/energy_model.py

```diff
@@ -0,0 +1,572 @@
+        spent not-idle in the range 0.0 - 1.0.
+
+        If combine=False, return idle and active power as separate components.
+        """
+        power = 0
+        ret = {}
+
+        assert all(0.0 <= a <= 1.0 for a in cpu_active_time)
+
+        for node in self.root.iter_nodes():
+            if not node.active_states or not node.idle_states:
+                continue
+
+            cpus = tuple(node.cpus)
+            # For now we assume topology nodes with energy models do not overlap
+            # with frequency domains
+            freq = freqs[cpus[0]]
+            assert all(freqs[c] == freq for c in cpus[1:])
+
+            # The active time of a node is estimated as the max of the active
+            # times of its children.
+            # This works great for the synthetic periodic workloads we use in
+            # LISA (where all threads wake up at the same time) but is probably
+            # no good for real workloads.
+            active_time = max(cpu_active_time[c] for c in cpus)
+            active_power = node.active_states[freq].power * active_time
+
+            _idle_power = max(node.idle_states[idle_states[c]] for c in cpus)
+            idle_power = _idle_power * (1 - active_time)
+
+            if combine:
+                ret[cpus] = active_power + idle_power
+            else:
+                ret[cpus] = {}
+                ret[cpus]["active"] = active_power
+                ret[cpus]["idle"] = idle_power
+
+        return ret
+
+    def estimate_from_cpu_util(self, util_distrib, freqs=None, idle_states=None):
+        """
+        Estimate the energy usage of the system under a utilization distribution
+
+        Optionally also take freqs; a list of frequencies at which each CPU is
+        assumed to run, and idle_states, the idle states that each CPU can enter
+        between activations. If not provided, they will be estimated assuming an
+        ideal selection system (i.e. perfect cpufreq & cpuidle governors).
+
+        :param util_distrib: Utilization distribution, see
+                             :ref:`util_distribs <util-distrib>`
+        :param freqs: List of CPU frequencies. Got from :meth:`guess_freqs` by
+                      default.
+        :param idle_states: List of CPU frequencies. Got from
+                            :meth:`guess_idle_states` by default.
+
+        :returns: Dict with power in bogo-Watts (bW), with contributions from
+                  each system component keyed with a tuple of the CPUs
+                  comprising that component (i.e. :attr:EnergyModelNode.cpus)
+
+                  ::
+
+                    {
+                        (0,)    : 10,
+                        (1,)    : 10,
+                        (1, 2)  : 5,
+                    }
+
+                  This represents CPUs 0 and 1 each using 10bW and their shared
+                  resources using 5bW for a total of 25bW.
+        """
+        if len(util_distrib) != len(self.cpus):
```

Comment:
> What about `cpus_utilizations`?
Hmm, yeah I think that would be a sensible name.

> it seems that you want to support the concept of variable workloads using "a series of util_distrib" ... but from this check this seems not to be supported right now, sin't it?
The implication of using a series of `util_distrib`s was that you'd call this function for each of them; that's what is done in generic.py (see `get_power_df`). Perhaps I can reword to make that clearer.